# Comparison of two cities: Toronto and New York #

### Introduction ###

New York City is the most populous city in the Unites States with a population of over 8 million people. It is also one of the most popular cities in the United States and a centre stage for many cultural events. In the same way Toronto is the most populous city in Canada wit a population of over 2 million people. Toronto is deemed as an international centre of bussiness and culture and is recognized as one of the most multicultural and cosmopolitan cities in the world. 

Two major cities of two major countries. An analysis comparing this two major cities would be of interest for people looking to setle in one of this cities. To have a vibrant and culturally stimulating life.

### Data ###

The data obtained comes from different sources. First, For the analysis, information on the list of neighbourhoods and the coordinates is needed to obtain information on the various venues present in each neibourhood and each city.

Initially, for Toronto, a table with neighbourhoods and postcodes was obtained from wikipedia. The table was cleaned to eliminate rows with 'not assigned' burough and for rows with 'not assigned' neighbours, the name of the burough was used instead.There was the instance of more than one neighbour per postcode. In this case the rows where merged to create one row with one postcode and the name of both neibourhoods separated by a comma.

We obtain the coordinates from http://cocl.us/Geospatial_data . The file contained the postcodes and their respective coordinates. We then merged both tables.

For New York City,

In [1]:
#importing libraries
import pandas as pd
import numpy as np
import requests
from bs4 import BeautifulSoup
from pandas.io.json import json_normalize
import json

In [2]:
# getting data and setting beautifulsoup
webseitdata=requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text
soup = BeautifulSoup(webseitdata,'html.parser')
# create lists to store data
postalCodeList = []
boroughList = []
neighborhoodList = []
#The table is available at class
for row in soup.find('table').find_all('tr'):
    cells = row.find_all('td')
    if (len(cells)>0):
        postalCodeList.append(cells[0].text)
        boroughList.append(cells[1].text)
        neighborhoodList.append(cells[2].text)
postalCodeList=list(map(lambda x:x.rstrip("\n"),postalCodeList))
boroughList=list(map(lambda x:x.rstrip("\n"),boroughList))
neighborhoodList=list(map(lambda x:x.rstrip("\n"),neighborhoodList))
df=pd.DataFrame({"PostalCode":postalCodeList, "Borough":boroughList,"Neighborhood":neighborhoodList})
df.head()

PostalCode           Borough               Neighborhood
0        M1A      Not assigned               Not assigned
1        M2A      Not assigned               Not assigned
2        M3A        North York                  Parkwoods
3        M4A        North York           Victoria Village
4        M5A  Downtown Toronto  Regent Park, Harbourfront

Here we process the table to deal with missing values as explained above.

In [3]:
#missing Burough data
df_clean=df[df.Borough!="Not assigned"].reset_index(drop=True)
# dealing with more than one neighbour per postcode
df_clean = df_clean.groupby(["PostalCode", "Borough"], as_index=False).agg(lambda x: ", ".join(x))
#missing neighbour data
df_clean["Neighborhood"]=df_clean["Neighborhood"].replace("Not assigned",df_clean["Borough"])
df_clean.head()

PostalCode      Borough                            Neighborhood
0        M1B  Scarborough                          Malvern, Rouge
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2        M1E  Scarborough       Guildwood, Morningside, West Hill
3        M1G  Scarborough                                  Woburn
4        M1H  Scarborough                               Cedarbrae

In [4]:
df_clean.shape

(103, 3)

We then get the coordinates for each postcode to use later to get venues information.

In [5]:
#get the table with the long and lat coordinates using the table at http://cocl.us/Geospatial_data
can_geo=pd.read_csv('http://cocl.us/Geospatial_data')
can_geo.columns=['PostalCode', 'Latitude', 'Longitude']
# Merge the table of coordinates with the previous table of boroughs and neighbourhoods using the postal code
toronto_df=pd.merge(df_clean,can_geo,how='inner',on="PostalCode")
toronto_df.head()

PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Malvern, Rouge  43.806686   
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

For New York City, we obtained the data of the neighbourhoods and coordinates from https://cocl.us/new_york_dataset . The file is a json file that contains the borough, neibourhoods and coordinates of New York City.

In [6]:
#getting new york data
!wget -q -O 'newyork_data.json' https://cocl.us/new_york_dataset
with open('newyork_data.json') as json_data:
    newyork_data = json.load(json_data)
neighborhoods_data = newyork_data['features']
neighborhoods_data[0]

{'type': 'Feature',
 'id': 'nyu_2451_34572.1',
 'geometry': {'type': 'Point',
  'coordinates': [-73.84720052054902, 40.89470517661]},
 'geometry_name': 'geom',
 'properties': {'name': 'Wakefield',
  'stacked': 1,
  'annoline1': 'Wakefield',
  'annoline2': None,
  'annoline3': None,
  'annoangle': 0.0,
  'borough': 'Bronx',
  'bbox': [-73.84720052054902,
   40.89470517661,
   -73.84720052054902,
   40.89470517661]}}

In [7]:
# define the dataframe columns
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 

# instantiate the dataframe
new_york = pd.DataFrame(columns=column_names)
for data in neighborhoods_data:
    borough = neighborhood_name = data['properties']['borough'] 
    neighborhood_name = data['properties']['name']
        
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]
    
    new_york = new_york.append({'Borough': borough,
                                          'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)
new_york.head()
    


Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

In [8]:
print('The Toronto has {} boroughs and {} neighborhoods.'.format(
        len(toronto_df['Borough'].unique()),
        toronto_df.shape[0]
    )
)
print('The New York has {} boroughs and {} neighborhoods.'.format(
        len(new_york['Borough'].unique()),
        new_york.shape[0]
    )
)

The Toronto has 10 boroughs and 103 neighborhoods.
The New York has 5 boroughs and 306 neighborhoods.


In [13]:
from geopy.geocoders import Nominatim
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans
#!conda install -c conda-forge folium=0.5.0 --yes
import folium 

In [15]:
address = 'Toronto,ON'

geolocator = Nominatim(user_agent="ON")
location = geolocator.geocode(address)
latitudet = location.latitude
longitudet = location.longitude
print('The geographical coordinate of Toronto are {}, {}.'.format(latitudet, longitudet))

The geographical coordinate of Toronto are 43.6534817, -79.3839347.


In [16]:
# create map of Toronto using latitude and longitude values of boroughs and neibourhoods of Toronto
map_toronto = folium.Map(location=[latitudet, longitudet], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_df['Latitude'], toronto_df['Longitude'], toronto_df['Borough'], toronto_df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [19]:
address = 'New York City, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of New York City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of New York City are 40.7127281, -74.0060152.


In [20]:
# create map of new york using latitude and longitude values of boroughs and neibourhoods of Toronto
map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(new_york['Latitude'], new_york['Longitude'], new_york['Borough'], new_york['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

Using the coordinate information and the Foursquare API, we get the information on the venues for each neighbourhood in both Toronto and New York City.

In [21]:
#Foursquare credentials definition
CLIENT_ID = 'OYQSKVZSJUT4DHCFH0SGXQO5Z1AJ0PSHV0C5N4G3IJDT5BMH' # your Foursquare ID
CLIENT_SECRET = 'KYQ1KTNDYSDJWW2STTLPRLGR21VMH4ZBXYNRZ5TGK2M4TNHT' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

In [22]:
# Fuction that gets the venues within 500 metres from each neighbourhood
def getNearbyVenues(names, latitudes, longitudes, radius=500,LIMIT=100):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    return(nearby_venues)

In [23]:
toronto_venues = getNearbyVenues(names=toronto_df['Neighborhood'],
                                   latitudes=toronto_df['Latitude'],
                                   longitudes=toronto_df['Longitude']
                                  )

Malvern, Rouge
Rouge Hill, Port Union, Highland Creek
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
Kennedy Park, Ionview, East Birchmount Park
Golden Mile, Clairlea, Oakridge
Cliffside, Cliffcrest, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Wexford Heights, Scarborough Town Centre
Wexford, Maryvale
Agincourt
Clarks Corners, Tam O'Shanter, Sullivan
Milliken, Agincourt North, Steeles East, L'Amoreaux East
Steeles West, L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
York Mills, Silver Hills
Willowdale, Newtonbrook
Willowdale, Willowdale East
York Mills West
Willowdale, Willowdale West
Parkwoods
Don Mills
Don Mills
Bathurst Manor, Wilson Heights, Downsview North
Northwood Park, York University
Downsview
Downsview
Downsview
Downsview
Victoria Village
Parkview Hill, Woodbine Gardens
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto, Broadview North (Old East York)
The Danforth West, 

In [24]:
newyork_venues = getNearbyVenues(names=new_york['Neighborhood'],
                                   latitudes=new_york['Latitude'],
                                   longitudes=new_york['Longitude']
                                  )

Wakefield
Co-op City
Eastchester
Fieldston
Riverdale
Kingsbridge
Marble Hill
Woodlawn
Norwood
Williamsbridge
Baychester
Pelham Parkway
City Island
Bedford Park
University Heights
Morris Heights
Fordham
East Tremont
West Farms
High  Bridge
Melrose
Mott Haven
Port Morris
Longwood
Hunts Point
Morrisania
Soundview
Clason Point
Throgs Neck
Country Club
Parkchester
Westchester Square
Van Nest
Morris Park
Belmont
Spuyten Duyvil
North Riverdale
Pelham Bay
Schuylerville
Edgewater Park
Castle Hill
Olinville
Pelham Gardens
Concourse
Unionport
Edenwald
Bay Ridge
Bensonhurst
Sunset Park
Greenpoint
Gravesend
Brighton Beach
Sheepshead Bay
Manhattan Terrace
Flatbush
Crown Heights
East Flatbush
Kensington
Windsor Terrace
Prospect Heights
Brownsville
Williamsburg
Bushwick
Bedford Stuyvesant
Brooklyn Heights
Cobble Hill
Carroll Gardens
Red Hook
Gowanus
Fort Greene
Park Slope
Cypress Hills
East New York
Starrett City
Canarsie
Flatlands
Mill Island
Manhattan Beach
Coney Island
Bath Beach
Borough Park
Dyker

### Methodology

Once we have obtained the neighbourhood and venues data, we can move on analyzing the information. Because the sheer number of the information, we will calculate the frequency of types of venues and Kmeans to cluster the data and compare the clusters between the cities.

### Analysis

In [26]:
print('{} Toronto venues were returned by Foursquare.'.format(toronto_venues.shape[0]))
print('{} New Yorkvenues were returned by Foursquare.'.format(newyork_venues.shape[0]))

2154 Toronto venues were returned by Foursquare.
10097 New Yorkvenues were returned by Foursquare.


In order to use KMeans, we expand the Venue Category column. Using get_dummies, we create new columns with the categories on 'Venues Categories.Then group them by Neighbourhood to get the frequency.  

In [27]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]
#print(toronto_onehot['Neighborhood'] )
toronto_onehot.head()

Yoga Studio  Accessories Store  Afghan Restaurant  Airport  \
0            0                  0                  0        0   
1            0                  0                  0        0   
2            0                  0                  0        0   
3            0                  0                  0        0   
4            0                  0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Toy / Game Store  Trail  \
0                 0                    0  ...                 0      0   
1                 0                    0  ...                 0      0   
2                 0                    0  ...                 0      0   
3                 0                    0  ...                 0      0   
4                 0                    0  ...                 0      0   

   Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0              0                              0                 0   
1              0                              0                 0   
2              0                              0                 0   
3              0                              0                 0   
4              0                              0                 0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  
0              0  
1              0  
2              0  
3              0  
4              0  

[5 rows x 267 columns]

In [28]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Yoga Studio  \
0                                        Agincourt          0.0   
1                           Alderwood, Long Branch          0.0   
2  Bathurst Manor, Wilson Heights, Downsview North          0.0   
3                                  Bayview Village          0.0   
4                Bedford Park, Lawrence Manor East          0.0   

   Accessories Store  Afghan Restaurant  Airport  Airport Food Court  \
0                0.0                0.0      0.0                 0.0   
1                0.0                0.0      0.0                 0.0   
2                0.0                0.0      0.0                 0.0   
3                0.0                0.0      0.0                 0.0   
4                0.0                0.0      0.0                 0.0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  ...  \
0           0.0             0.0              0.0               0.0  ...   
1           0.0             0.0              0.0               0.0  ...   
2           0.0             0.0              0.0               0.0  ...   
3           0.0             0.0              0.0               0.0  ...   
4           0.0             0.0              0.0               0.0  ...   

   Toy / Game Store  Trail  Train Station  Vegetarian / Vegan Restaurant  \
0          0.000000    0.0            0.0                            0.0   
1          0.000000    0.0            0.0                            0.0   
2          0.000000    0.0            0.0                            0.0   
3          0.000000    0.0            0.0                            0.0   
4          0.037037    0.0            0.0                            0.0   

   Video Game Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  \
0               0.0                    0.0              0.0       0.0   
1               0.0                    0.0              0.0       0.0   
2               0.0                    0.0              0.0       0.0   
3               0.0                    0.0              0.0       0.0   
4               0.0                    0.0              0.0       0.0   

   Wings Joint  Women's Store  
0          0.0            0.0  
1          0.0            0.0  
2          0.0            0.0  
3          0.0            0.0  
4          0.0            0.0  

[5 rows x 267 columns]

In [29]:
# one hot encoding
newyork_onehot = pd.get_dummies(newyork_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
newyork_onehot['Neighborhood'] = newyork_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [newyork_onehot.columns[-1]] + list(newyork_onehot.columns[:-1])
newyork_onehot = newyork_onehot[fixed_columns]
#print(toronto_onehot['Neighborhood'] )
newyork_grouped = newyork_onehot.groupby('Neighborhood').mean().reset_index()
newyork_grouped.head()

Neighborhood  Yoga Studio  Accessories Store  Adult Boutique  \
0       Allerton          0.0                0.0             0.0   
1       Annadale          0.0                0.0             0.0   
2  Arden Heights          0.0                0.0             0.0   
3      Arlington          0.0                0.0             0.0   
4       Arrochar          0.0                0.0             0.0   

   Afghan Restaurant  African Restaurant  Airport Terminal  \
0                0.0                 0.0               0.0   
1                0.0                 0.0               0.0   
2                0.0                 0.0               0.0   
3                0.0                 0.0               0.0   
4                0.0                 0.0               0.0   

   American Restaurant  Antique Shop  Aquarium  ...  Volleyball Court  \
0             0.000000           0.0       0.0  ...               0.0   
1             0.142857           0.0       0.0  ...               0.0   
2             0.000000           0.0       0.0  ...               0.0   
3             0.000000           0.0       0.0  ...               0.0   
4             0.000000           0.0       0.0  ...               0.0   

   Warehouse Store  Waste Facility  Waterfront  Weight Loss Center  \
0              0.0             0.0         0.0                 0.0   
1              0.0             0.0         0.0                 0.0   
2              0.0             0.0         0.0                 0.0   
3              0.0             0.0         0.0                 0.0   
4              0.0             0.0         0.0                 0.0   

   Whisky Bar  Wine Bar  Wine Shop  Wings Joint  Women's Store  
0         0.0       0.0        0.0          0.0            0.0  
1         0.0       0.0        0.0          0.0            0.0  
2         0.0       0.0        0.0          0.0            0.0  
3         0.0       0.0        0.0          0.0            0.0  
4         0.0       0.0        0.0          0.0            0.0  

[5 rows x 429 columns]

This is a function to sort the venues in descending order.

In [30]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Select the top ten common venues and ordering them.

In [31]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
toronto_venues_sorted = pd.DataFrame(columns=columns)
toronto_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
   toronto_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

toronto_venues_sorted

Neighborhood 1st Most Common Venue  \
0                                           Agincourt        Breakfast Spot   
1                              Alderwood, Long Branch           Pizza Place   
2     Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                     Bayview Village                  Café   
4                   Bedford Park, Lawrence Manor East            Restaurant   
5                                         Berczy Park           Coffee Shop   
6                         Birch Cliff, Cliffside West       College Stadium   
7        Brockton, Parkdale Village, Exhibition Place                  Café   
8   Business reply mail Processing Centre, South C...    Light Rail Station   
9   CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
10                                Caledonia-Fairbanks                  Park   
11              Canada Post Gateway Processing Centre                 Hotel   
12                                          Cedarbrae   Fried Chicken Joint   
13                                 Central Bay Street           Coffee Shop   
14                                           Christie         Grocery Store   
15                               Church and Wellesley           Coffee Shop   
16            Clarks Corners, Tam O'Shanter, Sullivan           Pizza Place   
17    Cliffside, Cliffcrest, Scarborough Village West   American Restaurant   
18                     Commerce Court, Victoria Hotel           Coffee Shop   
19                                         Davisville          Dessert Shop   
20                                   Davisville North                 Hotel   
21   Del Ray, Mount Dennis, Keelsdale and Silverthorn           Coffee Shop   
22                                          Don Mills                   Gym   
23  Dorset Park, Wexford Heights, Scarborough Town...     Indian Restaurant   
24                                          Downsview         Grocery Store   
25                       Dufferin, Dovercourt Village              Pharmacy   
26      East Toronto, Broadview North (Old East York)                  Park   
27  Eringate, Bloordale Gardens, Old Burnhamthorpe...             Pet Store   
28                       Fairview, Henry Farm, Oriole        Clothing Store   
29             First Canadian Place, Underground city           Coffee Shop   
..                                                ...                   ...   
65                                           Rosedale                  Park   
66                                           Roselawn           Music Venue   
67             Rouge Hill, Port Union, Highland Creek                   Bar   
68                                 Runnymede, Swansea           Coffee Shop   
69                      Runnymede, The Junction North         Grocery Store   
70                                Scarborough Village            Playground   
71  South Steeles, Silverstone, Humbergate, Jamest...         Grocery Store   
72                                     St. James Town                  Café   
73                        St. James Town, Cabbagetown                  Café   
74                      Steeles West, L'Amoreaux West  Fast Food Restaurant   
75                                     Stn A PO Boxes           Coffee Shop   
76                                    Studio District                  Café   
77  Summerhill West, Rathnelly, South Hill, Forest...                   Pub   
78                The Annex, North Midtown, Yorkville                  Café   
79                                        The Beaches      Asian Restaurant   
80                       The Danforth West, Riverdale      Greek Restaurant   
81      The Kingsway, Montgomery Road, Old Mill North                  Park   
82                                   Thorncliffe Park        Sandwich Place   
83           Toronto Dominion Centre, Design Exchange           Coffee Shop   
84          

In [32]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
newyork_venues_sorted = pd.DataFrame(columns=columns)
newyork_venues_sorted['Neighborhood'] = newyork_grouped['Neighborhood']

for ind in np.arange(newyork_grouped.shape[0]):
   newyork_venues_sorted.iloc[ind, 1:] = return_most_common_venues(newyork_grouped.iloc[ind, :], num_top_venues)

newyork_venues_sorted

Neighborhood 1st Most Common Venue       2nd Most Common Venue  \
0              Allerton           Pizza Place               Deli / Bodega   
1              Annadale           Pizza Place         American Restaurant   
2         Arden Heights           Pizza Place                    Pharmacy   
3             Arlington              Bus Stop                Intersection   
4              Arrochar              Bus Stop          Italian Restaurant   
5               Arverne             Surf Spot               Metro Station   
6               Astoria                   Bar   Middle Eastern Restaurant   
7       Astoria Heights    Italian Restaurant            Business Service   
8            Auburndale    Italian Restaurant           Mobile Phone Shop   
9            Bath Beach    Chinese Restaurant                    Pharmacy   
10    Battery Park City                  Park                       Hotel   
11            Bay Ridge                   Spa                 Pizza Place   
12          Bay Terrace        Clothing Store                 Supermarket   
13           Baychester            Donut Shop         Fried Chicken Joint   
14              Bayside                   Bar                         Pub   
15            Bayswater            Playground  Construction & Landscaping   
16         Bedford Park    Chinese Restaurant                       Diner   
17   Bedford Stuyvesant           Coffee Shop                        Café   
18           Beechhurst           Yoga Studio                 Supermarket   
19             Bellaire           Bus Station                 Coffee Shop   
20         Belle Harbor                 Beach                         Pub   
21            Bellerose         Deli / Bodega                 Pizza Place   
22              Belmont    Italian Restaurant                 Pizza Place   
23          Bensonhurst            Donut Shop          Italian Restaurant   
24         Bergen Beach       Harbor / Marina                Hockey Field   
25           Blissville            Donut Shop               Deli / Bodega   
26           Bloomfield     Recreation Center                Burger Joint   
27          Boerum Hill          Dance Studio                 Coffee Shop   
28         Borough Park                  Bank                    Pharmacy   
29         Breezy Point                 Beach                       Trail   
..                  ...                   ...                         ...   
271              Travis                 Hotel               Bowling Alley   
272             Tribeca                  Park          Italian Restaurant   
273          Tudor City                  Park                        Café   
274          Turtle Bay      Sushi Restaurant          Italian Restaurant   
275           Unionport            Donut Shop               Deli / Bodega   
276  University Heights           Pizza Place          Chinese Restaurant   
277     Upper East Side    Italian Restaurant                 Coffee Shop   
278     Upper West Side    Italian Restaurant                         Bar   
279              Utopia         Deli / Bodega                  Playground   
280            Van Nest         Deli / Bodega                 Pizza Place   
281        Vinegar Hill            Food Truck                 Coffee Shop   
282           Wakefield              Pharmacy                 Pizza Place   
283  Washington Heights                  Café                      Bakery   
284          Weeksville        Discount Store                Cocktail Bar   
285       West Brighton           Coffee Shop                 Music Store   
286          West Farms           Bus Station                        Park   
287        West Village    Italian Restaurant     New American Restaurant   
288  Westchester Square  Fast Food Restaurant                    Pharmacy   
289         Westerleigh     Convenience Store            Sushi Restaurant   
290          Whitestone       Bubble Tea Shop                Dance Studio   
291    

Using Kmeans to identify 5 clusters within our neighbourhoods.

In [33]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)
# run k-means clustering
kmeanst = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

newyork_grouped_clustering = newyork_grouped.drop('Neighborhood', 1)
# run k-means clustering
kmeansn = KMeans(n_clusters=kclusters, random_state=0).fit(newyork_grouped_clustering)
# check cluster labels generated for each row in the dataframe
kmeansn.labels_[0:10] 

array([4, 4, 4, 3, 3, 4, 4, 4, 4, 4], dtype=int32)

In [34]:
# add clustering labels
toronto_venues_sorted.insert(0, 'ClusterLabels', kmeanst.labels_)
newyork_venues_sorted.insert(0, 'ClusterLabels', kmeansn.labels_)

In [35]:

toronto_merged = toronto_df
toronto_merged = toronto_merged.drop([16,0], axis=0)

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(toronto_venues_sorted.set_index('Neighborhood'), on='Neighborhood')
toronto_merged.dropna(inplace=True)
toronto_merged

PostalCode           Borough  \
1          M1C       Scarborough   
2          M1E       Scarborough   
3          M1G       Scarborough   
4          M1H       Scarborough   
5          M1J       Scarborough   
6          M1K       Scarborough   
7          M1L       Scarborough   
8          M1M       Scarborough   
9          M1N       Scarborough   
10         M1P       Scarborough   
11         M1R       Scarborough   
12         M1S       Scarborough   
13         M1T       Scarborough   
14         M1V       Scarborough   
15         M1W       Scarborough   
17         M2H        North York   
18         M2J        North York   
19         M2K        North York   
21         M2M        North York   
22         M2N        North York   
23         M2P        North York   
24         M2R        North York   
25         M3A        North York   
26         M3B        North York   
27         M3C        North York   
28         M3H        North York   
29         M3J        North York   
30         M3K        North York   
31         M3L        North York   
32         M3M        North York   
..         ...               ...   
71         M6A        North York   
72         M6B        North York   
73         M6C              York   
74         M6E              York   
75         M6G  Downtown Toronto   
76         M6H      West Toronto   
77         M6J      West Toronto   
78         M6K      West Toronto   
79         M6L        North York   
80         M6M              York   
81         M6N              York   
82         M6P      West Toronto   
83         M6R      West Toronto   
84         M6S      West Toronto   
85         M7A  Downtown Toronto   
86         M7R       Mississauga   
87         M7Y      East Toronto   
88         M8V         Etobicoke   
89         M8W         Etobicoke   
90         M8X         Etobicoke   
91         M8Y         Etobicoke   
92         M8Z         Etobicoke   
95         M9C         Etobicoke   
96         M9L        North York   
97         M9M        North York   
98         M9N              York   
99         M9P         Etobicoke   
100        M9R         Etobicoke   
101        M9V         Etobicoke   
102        M9W         Etobicoke   

                                          Neighborhood   Latitude  Longitude  \
1               Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497   
2                    Guildwood, Morningside, West Hill  43.763573 -79.188711   
3                                               Woburn  43.770992 -79.216917   
4                                            Cedarbrae  43.773136 -79.239476   
5                                  Scarborough Village  43.744734 -79.239476   
6          Kennedy Park, Ionview, East Birchmount Park  43.727929 -79.262029   
7                      Golden Mile, Clairlea, Oakridge  43.711112 -79.284577   
8      Cliffside, Cliffcrest, Scarborough Village West  43.716316 -79.239476   
9                          Birch Cliff, Cliffside West  43.692657 -79.264848   
10   Dorset Park, Wexford Heights, Scarborough Town...  43.757410 -79.273304   
11                                   Wexford, Maryvale  43.750072 -79.295849   
12                                           Agincourt  43.794200 -79.262029   
13             Clarks Corners, Tam O'Shanter, Sullivan  43.781638 -79.304302   
14   Milliken, Agincourt North, Steeles East, L'Amo...  43.815252 -79.284577   
15                       Steeles West, L'Amoreaux West  43.799525 -79.318389   
17                                   Hillcrest Village  43.803762 -79.363452   
18                        Fairview, Henry Farm, Oriole  43.778517 -79.346556   
19                                     Bayview Village  43.786947 -79.385975   
21                             Willowdale, Newtonbrook  43.789053 -79.408493   
22                         Willowdale, Willowdale East  43.770120 -79.408493   
23                                     York Mills West  43.752758 -79.400049   
24         

In [36]:

newyork_merged = new_york
newyork_merged = newyork_merged.drop([16,0], axis=0)

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
newyork_merged = newyork_merged.join(newyork_venues_sorted.set_index('Neighborhood'), on='Neighborhood')
newyork_merged.dropna(inplace=True)
newyork_merged

Borough         Neighborhood   Latitude  Longitude  ClusterLabels  \
1            Bronx           Co-op City  40.874294 -73.829939            4.0   
2            Bronx          Eastchester  40.887556 -73.827806            3.0   
3            Bronx            Fieldston  40.895437 -73.905643            4.0   
4            Bronx            Riverdale  40.890834 -73.912585            4.0   
5            Bronx          Kingsbridge  40.881687 -73.902818            4.0   
6        Manhattan          Marble Hill  40.876551 -73.910660            4.0   
7            Bronx             Woodlawn  40.898273 -73.867315            3.0   
8            Bronx              Norwood  40.877224 -73.879391            4.0   
9            Bronx       Williamsbridge  40.881039 -73.857446            4.0   
10           Bronx           Baychester  40.866858 -73.835798            4.0   
11           Bronx       Pelham Parkway  40.857413 -73.854756            4.0   
12           Bronx          City Island  40.847247 -73.786488            4.0   
13           Bronx         Bedford Park  40.870185 -73.885512            4.0   
14           Bronx   University Heights  40.855727 -73.910416            4.0   
15           Bronx       Morris Heights  40.847898 -73.919672            4.0   
17           Bronx         East Tremont  40.842696 -73.887356            4.0   
18           Bronx           West Farms  40.839475 -73.877745            4.0   
19           Bronx         High  Bridge  40.836623 -73.926102            4.0   
20           Bronx              Melrose  40.819754 -73.909422            4.0   
21           Bronx           Mott Haven  40.806239 -73.916100            4.0   
22           Bronx          Port Morris  40.801664 -73.913221            4.0   
23           Bronx             Longwood  40.815099 -73.895788            3.0   
24           Bronx          Hunts Point  40.809730 -73.883315            4.0   
25           Bronx           Morrisania  40.823592 -73.901506            4.0   
26           Bronx            Soundview  40.821012 -73.865746            4.0   
27           Bronx         Clason Point  40.806551 -73.854144            0.0   
28           Bronx          Throgs Neck  40.815109 -73.816350            3.0   
29           Bronx         Country Club  40.844246 -73.824099            4.0   
30           Bronx          Parkchester  40.837938 -73.856003            4.0   
31           Bronx   Westchester Square  40.840619 -73.842194            4.0   
..             ...                  ...        ...        ...            ...   
276      Manhattan             Flatiron  40.739673 -73.990947            4.0   
277         Queens    Sunnyside Gardens  40.745652 -73.918193            4.0   
278         Queens           Blissville  40.737251 -73.932442            3.0   
279       Brooklyn         Fulton Ferry  40.703281 -73.995508            4.0   
280       Brooklyn         Vinegar Hill  40.703321 -73.981116            4.0   
281       Brooklyn           Weeksville  40.675040 -73.930531            4.0   
282       Brooklyn    Broadway Junction  40.677861 -73.903317            4.0   
283       Brooklyn                Dumbo  40.703176 -73.988753            4.0   
284  Staten Island        Manor Heights  40.601810 -74.120594            3.0   
285  Staten Island          Willowbrook  40.603707 -74.132084            3.0   
286  Staten Island         Sandy Ground  40.541140 -74.217766            4.0   
287  Staten Island          Egbertville  40.579119 -74.127272            4.0   
288         Queens              Roxbury  40.567376 -73.892138            3.0   
289       Brooklyn            Homecrest  40.598525 -73.959185            4.0   
290         Queens       Middle Village  40.716415 -73.881143            4.0   
291  Staten Island         Prince's Bay  40.526264 -74.201526            4.0   
292  Staten Island      Lighthouse Hill  40.576506 -74.137927            4.0   
293  Staten Island      Richmond Valley  40.519541 -74.229571            4.0   
294        

In [43]:
# create map
toronto_merged['ClusterLabels']=toronto_merged['ClusterLabels'].astype(int)
mapt_clusters = folium.Map(location=[latitudet, longitudet], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['ClusterLabels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(mapt_clusters)
       
mapt_clusters

In [44]:
# create map
newyork_merged['ClusterLabels']=newyork_merged['ClusterLabels'].astype(int)
mapn_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(newyork_merged['Latitude'], newyork_merged['Longitude'], newyork_merged['Neighborhood'], newyork_merged['ClusterLabels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(mapn_clusters)
       
mapn_clusters

- Toronto's Clusters

In [46]:
toronto_merged.loc[toronto_merged['ClusterLabels'] == 0,toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Empty DataFrame
Columns: [Borough, ClusterLabels, 1st Most Common Venue, 2nd Most Common Venue, 3rd Most Common Venue, 4th Most Common Venue, 5th Most Common Venue, 6th Most Common Venue, 7th Most Common Venue, 8th Most Common Venue, 9th Most Common Venue, 10th Most Common Venue]
Index: []

In [47]:
toronto_merged.loc[toronto_merged['ClusterLabels'] == 1,toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  ClusterLabels 1st Most Common Venue  \
1         Scarborough              1                   Bar   
2         Scarborough              1     Electronics Store   
3         Scarborough              1           Coffee Shop   
4         Scarborough              1   Fried Chicken Joint   
5         Scarborough              1            Playground   
6         Scarborough              1            Hobby Shop   
7         Scarborough              1              Bus Line   
8         Scarborough              1   American Restaurant   
9         Scarborough              1       College Stadium   
10        Scarborough              1     Indian Restaurant   
11        Scarborough              1           Auto Garage   
12        Scarborough              1        Breakfast Spot   
13        Scarborough              1           Pizza Place   
15        Scarborough              1  Fast Food Restaurant   
17         North York              1    Athletics & Sports   
18         North York              1        Clothing Store   
19         North York              1                  Café   
22         North York              1      Ramen Restaurant   
24         North York              1         Grocery Store   
26         North York              1                   Gym   
27         North York              1                   Gym   
28         North York              1           Coffee Shop   
29         North York              1                   Bar   
30         North York              1         Grocery Store   
31         North York              1         Grocery Store   
32         North York              1         Grocery Store   
33         North York              1         Grocery Store   
34         North York              1          Hockey Arena   
35          East York              1           Pizza Place   
36          East York              1           Curling Ice   
..                ...            ...                   ...   
61   Downtown Toronto              1           Coffee Shop   
62         North York              1            Restaurant   
64    Central Toronto              1                  Park   
65    Central Toronto              1                  Café   
66   Downtown Toronto              1                  Café   
67   Downtown Toronto              1                  Café   
68   Downtown Toronto              1       Airport Service   
69   Downtown Toronto              1           Coffee Shop   
70   Downtown Toronto              1           Coffee Shop   
71         North York              1        Clothing Store   
73               York              1                 Field   
75   Downtown Toronto              1         Grocery Store   
76       West Toronto              1              Pharmacy   
77       West Toronto              1                   Bar   
78       West Toronto              1                  Café   
79         North York              1      Basketball Court   
80               York              1           Coffee Shop   
81               York              1         Grocery Store   
82       West Toronto              1                  Café   
83       West Toronto              1             Gift Shop   
84       West Toronto              1           Coffee Shop   
85   Downtown Toronto              1           Coffee Shop   
86        Mississauga              1                 Hotel   
87       East Toronto              1    Light Rail Station   
88          Etobicoke              1                  Café   
92          Etobicoke              1         Grocery Store   
95          Etobicoke              1             Pet Store   
97         North York              1        Baseball Field   
101         Etobicoke              1         Grocery Store   
102         Etobicoke              1   Rental Car Location   

             2nd Most Common Venue   3rd Most Common Venue  \
1                     Home Service           Women's Store   
2                       Restaurant     Rental

In [48]:
toronto_merged.loc[toronto_merged['ClusterLabels'] == 2,toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  ClusterLabels 1st Most Common Venue  \
63  Central Toronto              2           Music Venue   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
63                Garden         Women's Store        Discount Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
63      Department Store          Dessert Shop    Dim Sum Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
63                 Diner   Distribution Center           Dance Studio

In [49]:
toronto_merged.loc[toronto_merged['ClusterLabels'] == 3,toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  ClusterLabels 1st Most Common Venue  \
14        Scarborough              3                  Park   
21         North York              3                  Park   
23         North York              3                  Park   
25         North York              3                  Park   
40          East York              3                  Park   
44    Central Toronto              3                  Park   
50   Downtown Toronto              3                  Park   
74               York              3                  Park   
90          Etobicoke              3                  Park   
91          Etobicoke              3                  Park   
98               York              3     Convenience Store   
100         Etobicoke              3                  Park   

    2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
14             Playground   Distribution Center                Deli / Bodega   
21          Women's Store               Dog Run             Department Store   
23      Convenience Store         Women's Store                 Dessert Shop   
25      Food & Drink Shop         Women's Store          Distribution Center   
40      Convenience Store         Women's Store                 Dessert Shop   
44               Bus Line           Swim School               Discount Store   
50             Playground                 Trail               Discount Store   
74                   Pool         Women's Store       Gluten-free Restaurant   
90                   Pool                 River           College Rec Center   
91          Deli / Bodega        Baseball Field                Women's Store   
98                   Park         Women's Store                 Dessert Shop   
100        Sandwich Place          Dance Studio  Eastern European Restaurant   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
14       Department Store          Dessert Shop    Dim Sum Restaurant   
21           Dessert Shop    Dim Sum Restaurant                 Diner   
23     Dim Sum Restaurant                 Diner        Discount Store   
25       Department Store          Dessert Shop    Dim Sum Restaurant   
40     Dim Sum Restaurant                 Diner        Discount Store   
44       Department Store          Dessert Shop    Dim Sum Restaurant   
50           Dance Studio         Deli / Bodega      Department Store   
74              Gift Shop   Dumpling Restaurant             Drugstore   
90            Curling Ice   Dumpling Restaurant             Drugstore   
91                Dog Run          Dessert Shop    Dim Sum Restaurant   
98     Dim Sum Restaurant                 Diner        Discount Store   
100   Dumpling Restaurant             Drugstore            Donut Shop   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
14                  Diner        Discount Store                Dog Run  
21         Discount Store   Distribution Center       Doner Restaurant  
23    Distribution Center               Dog Run       Doner Restaurant  
25                  Diner        Discount Store                Dog Run  
40    Distribution Center               Dog Run       Doner Restaurant  
44                  Diner   Distribution Center   Ethiopian Restaurant  
50           Dessert Shop    Dim Sum Restaurant                  Diner  
74             Donut Shop      Doner Restaurant                Dog Run  
90             Donut Shop      Doner Restaurant                Dog Run  
91                  Diner        Discount Store    Distribution Center  
98    Distribution Center               Dog Run       Doner Restaurant  
100      Doner Restaurant               Dog Run    Distribution Center

In [50]:
toronto_merged.loc[toronto_merged['ClusterLabels'] == 4,toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  ClusterLabels 1st Most Common Venue 2nd Most Common Venue  \
72  North York              4   Japanese Restaurant           Pizza Place   
89   Etobicoke              4           Pizza Place           Coffee Shop   
96  North York              4                   Gym           Pizza Place   
99   Etobicoke              4           Pizza Place        Discount Store   

   3rd Most Common Venue      4th Most Common Venue 5th Most Common Venue  \
72                Bakery                        Pub         Women's Store   
89                   Gym               Skating Rink    Athletics & Sports   
96         Women's Store                    Dog Run      Department Store   
99           Coffee Shop  Middle Eastern Restaurant          Intersection   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
72          Dessert Shop    Dim Sum Restaurant                 Diner   
89                   Pub        Sandwich Place         Women's Store   
96          Dessert Shop    Dim Sum Restaurant                 Diner   
99        Sandwich Place    Chinese Restaurant         Women's Store   

   9th Most Common Venue 10th Most Common Venue  
72        Discount Store    Distribution Center  
89                 Diner       Department Store  
96        Discount Store    Distribution Center  
99          Dessert Shop     Dim Sum Restaurant

- New York City's Clusters

In [51]:
newyork_merged.loc[newyork_merged['ClusterLabels'] == 0,newyork_merged.columns[[1] + list(range(5, newyork_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
27   Clason Point                  Park     Convenience Store   
188     Laurelton  Caribbean Restaurant                  Park   
192    Somerville                  Park         Women's Store   
203     Todt Hill                  Park         Women's Store   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
27          Grocery Store                  Pool              Bus Stop   
188         Train Station         Women's Store                 Field   
192    English Restaurant  Ethiopian Restaurant         Event Service   
203    English Restaurant  Ethiopian Restaurant         Event Service   

    6th Most Common Venue      7th Most Common Venue 8th Most Common Venue  \
27          Boat or Ferry  South American Restaurant         Women's Store   
188         Event Service                Event Space               Exhibit   
192           Event Space                    Exhibit               Factory   
203           Event Space                    Exhibit               Factory   

    9th Most Common Venue 10th Most Common Venue  
27                  Field            Event Space  
188               Factory     Falafel Restaurant  
192    Falafel Restaurant                   Farm  
203    Falafel Restaurant                   Farm

In [52]:
newyork_merged.loc[newyork_merged['ClusterLabels'] == 1,newyork_merged.columns[[1] + list(range(5, newyork_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
172  Breezy Point                 Beach                 Trail   
179      Neponsit                 Beach                   Bar   
204   South Beach                 Beach                  Pier   
302       Hammels                 Beach            Shoe Store   

          3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
172         Monument / Landmark         Women's Store   Filipino Restaurant   
179  Financial or Legal Service         Event Service           Event Space   
204               Deli / Bodega              Bus Stop    Athletics & Sports   
302                       Diner            Food Truck           Bus Station   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
172         Event Service           Event Space               Exhibit   
179               Exhibit               Factory    Falafel Restaurant   
204         Women's Store   Filipino Restaurant           Event Space   
302              Bus Stop  Fast Food Restaurant  Gym / Fitness Center   

    9th Most Common Venue 10th Most Common Venue  
172               Factory     Falafel Restaurant  
179                  Farm         Farmers Market  
204               Exhibit                Factory  
302               Dog Run    Fried Chicken Joint

In [53]:
newyork_merged.loc[newyork_merged['ClusterLabels'] == 2,newyork_merged.columns[[1] + list(range(5, newyork_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
183  Jamaica Estates           Bus Station               Dog Run   
202      Grymes Hill               Dog Run         Women's Store   

    3rd Most Common Venue       4th Most Common Venue 5th Most Common Venue  \
183         Women's Store  Financial or Legal Service         Event Service   
202   Filipino Restaurant        Ethiopian Restaurant         Event Service   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
183           Event Space               Exhibit               Factory   
202           Event Space               Exhibit               Factory   

    9th Most Common Venue 10th Most Common Venue  
183    Falafel Restaurant                   Farm  
202    Falafel Restaurant                   Farm

In [55]:
newyork_merged.loc[newyork_merged['ClusterLabels'] == 3,newyork_merged.columns[[1] + list(range(5, newyork_merged.shape[1]))]]

Neighborhood 1st Most Common Venue   2nd Most Common Venue  \
2         Eastchester  Caribbean Restaurant             Bus Station   
7            Woodlawn                   Pub           Deli / Bodega   
23           Longwood         Deli / Bodega           Grocery Store   
28        Throgs Neck    Italian Restaurant             Pizza Place   
32           Van Nest         Deli / Bodega             Pizza Place   
39     Edgewater Park    Italian Restaurant             Pizza Place   
41          Olinville  Caribbean Restaurant  Furniture / Home Store   
43          Concourse         Deli / Bodega           Grocery Store   
74           Canarsie    Chinese Restaurant           Deli / Bodega   
83        Marine Park        Baseball Field                    Park   
89         Ocean Hill         Deli / Bodega           Grocery Store   
144          Glendale           Pizza Place       Food & Drink Shop   
146         Woodhaven         Deli / Bodega                Pharmacy   
148  South Ozone Park         Deli / Bodega                    Park   
150        Whitestone       Bubble Tea Shop            Dance Studio   
156         Bellerose         Deli / Bodega             Pizza Place   
159         Briarwood            Playground     Arts & Crafts Store   
170      Far Rockaway         Deli / Bodega             Pizza Place   
171     Broad Channel              Dive Bar             Pizza Place   
190      Belle Harbor                 Beach                     Pub   
193        Brookville         Deli / Bodega           Women's Store   
194          Bellaire           Bus Station             Coffee Shop   
195      North Corona         Deli / Bodega    Gym / Fitness Center   
198      New Brighton              Bus Stop                    Park   
206  Mariner's Harbor    Italian Restaurant           Deli / Bodega   
210            Travis                 Hotel           Bowling Alley   
217       Tottenville    Mexican Restaurant      Italian Restaurant   
218     Tompkinsville         Deli / Bodega             Pizza Place   
227         Arlington              Bus Stop            Intersection   
228          Arrochar              Bus Stop      Italian Restaurant   
232     Midland Beach                 Beach      Russian Restaurant   
234    New Dorp Beach    Italian Restaurant           Deli / Bodega   
236          Huguenot        Ice Cream Shop           Deli / Bodega   
252       Shore Acres    Italian Restaurant                Bus Stop   
254           Concord    Chinese Restaurant     Peruvian Restaurant   
258          Elm Park    Italian Restaurant                Bus Stop   
261   Paerdegat Basin      Asian Restaurant         Harbor / Marina   
264            Utopia         Deli / Bodega              Playground   
278        Blissville            Donut Shop           Deli / Bodega   
284     Manor Heights            Bagel Shop              Donut Shop   
285       Willowbrook              Bus Stop      Chinese Restaurant   
288           Roxbury        Baseball Field               Irish Pub   
295     Highland Park         Deli / Bodega             Pizza Place   
296           Madison            Bagel Shop           Deli / Bodega   

          3rd Most Common Venue            4th Most Common Venue  \
2                 Deli / Bodega                            Diner   
7                   Pizza Place                       Playground   
23                        Diner             Fast Food Restaurant   
28                   Sports Bar                Mobile Phone Shop   
32                  Bus Station                           Bakery   
39                  Coffee Shop                    Deli / Bodega   
41             Basketball Court                    Deli / Bodega   
43           Chinese Restaurant                      Pizza Place   
74                     Bus Line                             Food   
83                  Pizza Place                              Gym   
89                   Playground  Southern / Soul Food Restaurant 

In [56]:
newyork_merged.loc[newyork_merged['ClusterLabels'] == 4,newyork_merged.columns[[1] + list(range(5, newyork_merged.shape[1]))]]

Neighborhood   1st Most Common Venue       2nd Most Common Venue  \
1             Co-op City    Fast Food Restaurant                  Bagel Shop   
3              Fieldston             Music Venue                       River   
4              Riverdale             Bus Station        Medical Supply Store   
5            Kingsbridge             Pizza Place                         Bar   
6            Marble Hill                     Gym                 Coffee Shop   
8                Norwood             Pizza Place          Chinese Restaurant   
9         Williamsbridge    Caribbean Restaurant                         Bar   
10            Baychester              Donut Shop         Fried Chicken Joint   
11        Pelham Parkway           Deli / Bodega          Italian Restaurant   
12           City Island  Thrift / Vintage Store          Seafood Restaurant   
13          Bedford Park      Chinese Restaurant                       Diner   
14    University Heights             Pizza Place          Chinese Restaurant   
15        Morris Heights                    Bank                 Bus Station   
17          East Tremont             Pizza Place              Cosmetics Shop   
18            West Farms             Bus Station                        Park   
19          High  Bridge      Chinese Restaurant                    Pharmacy   
20               Melrose             Pizza Place                    Pharmacy   
21            Mott Haven             Pizza Place                         Gym   
22           Port Morris                     Bar                  Distillery   
24           Hunts Point                    Bank                 Pizza Place   
25            Morrisania              Donut Shop                 Bus Station   
26             Soundview      Chinese Restaurant               Grocery Store   
29          Country Club          Sandwich Place                  Playground   
30           Parkchester             Supermarket                 Pizza Place   
31    Westchester Square    Fast Food Restaurant                    Pharmacy   
33           Morris Park             Pizza Place                Burger Joint   
34               Belmont      Italian Restaurant                 Pizza Place   
35        Spuyten Duyvil                    Park            Asian Restaurant   
36       North Riverdale             Pizza Place          Chinese Restaurant   
37            Pelham Bay      Italian Restaurant                        Bank   
..                   ...                     ...                         ...   
269           Mount Eden                Pharmacy               Deli / Bodega   
270           Mount Hope           Grocery Store                 Pizza Place   
271         Sutton Place      Italian Restaurant                 Coffee Shop   
272        Hunters Point                    Café          Italian Restaurant   
273           Turtle Bay        Sushi Restaurant          Italian Restaurant   
274           Tudor City                    Park                        Café   
275      Stuyvesant Town           Boat or Ferry                        Park   
276             Flatiron      Italian Restaurant                         Spa   
277    Sunnyside Gardens                     Bar               Grocery Store   
279         Fulton Ferry                    Park         American Restaurant   
280         Vinegar Hill              Food Truck                 Coffee Shop   
281           Weeksville          Discount Store                Cocktail Bar   
282    Broadway Junction     Fried Chicken Joint                  Donut Shop   
283                Dumbo                    Park                 Coffee Shop   
286         Sandy Ground               Racetrack                  Playground   
287          Egbertville      Italian Restaurant                  Bagel Shop   
289            Homecrest                    Bank                  Donut Shop   
290       Middle Village      Chinese Restaurant                Dessert Shop   
291         

### Results and discussion

Segmenting both cities using KMeans yielded 5 klusters per city. 

The clusters for Toronto:

- Cluster 1

There are no elements in this cluster
- Cluster 2

It is the largest cluster with 81 neighbourhoods out of 103. It shows a variety of restaurants from a variety of ethnicities and caffes.
- Cluster 3

There are not enough neighbourhood to form an opinion. Although similarly to cluster 1, there are a lot event related venues. Caffes seem to be very popular in the neighbourhoods of this region.
- Cluster 4 

The elements of this cluster have a major frequency of Park areas and Discount stores.
- Cluster 5 

The neighbourhoods of this cluster have all, at first or second place, Pizza place. They also have in similar women's store and distribution centre. 

The clusters for New York City :

- Cluster 1

Every neighbourhood boots in its top ten:  parks, event services, event space, and exhibits.
- Cluster 2

They all seem to be situated along beaches.
- Cluster 3

There are not enough neighbourhood to form an opinion. Although similarly to cluster 1, there are a lot event related venues.
- Cluster 4 

This cluster has a preponderance of Deli/Bodegas. It also shows a large variety of restaurants from many different ethnicities. A statement of the multicultural proprerty of New York City.
- Cluster 5 

It is the largest cluster with 249 neighbourhoods out of 306. It shows a variety of restaurants and leisure venues.

### Conclussion

Although both cities are the most populous cities of their country, New York City has 4 times more inhabitants that its counterpart, Toronto. This is reflected also in the number of neighbourhoods: 103 and 306 for Toronto and New York respectively. But it is in the venues where the difference is more seen; the 2154 of Toronto venues agains the 10097 of New York City. The number of venues in New York City is 5 times that of tge venues in Toronto.

The segmentation was very looped with one cluster having about 80% of the neibouroods and one or two clusters having 0-2 neighbourhoods. But some charateristics are still observable. Both cities show their multicultural cusine and the importance of leisure activities. The importance of the culture of the Deli/Bodegas in New York City is seen in cluster 4. And while there is a variety of cusines, Pizza places and caffes seem to be the most important in Toronto as seen in cluster 2 and 5.

We have outlined some similarities and differences of both this cities based on the types and frequency of venues. For people deciding to transfer, there are still others deciding factors as 